### PR TITLE
Replace javassist with org.ow2.asm

### LINKED ISSUE #841: Consider making javassist an optional dependency or move to separate module.
javassist is only used for `MainClassFinder`, so it should either be:
- made optional so that the dependency is only required if the user of Jib Core uses `MainClassFinder` (and other Java-specific code), or
- moved along with other Java-specific code to a separate module like `jib-core-extensions`

### PR DESCRIPTION
**Pull request by TadCordle**
All tests pass, but given how the main class finder is an important feature and historically has had many edge cases, we may want to do some extra manual testing on a big project.

This is about 1/7 the size of javassist, so I think we can say this fixes #841.

### CODE REVIEW

**Comment by TadCordle** [comment]
@briandealwis whoops, saw there was a 7.0 beta but missed that there was also a non-beta 7.0. I'll change it.

**Comment by TadCordle** [comment]
@coollog [ASM 7.0 seems to support Java 11.](https://asm.ow2.io/versions.html)

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java [hunk 1]
```diff
@@ -157,35 +179,21 @@ public Result find() {
...
+      ClassReader reader = new ClassReader(classFileInputStream);
+      reader.accept(mainClassVisitor, 0);
+      if (mainClassVisitor.visitedMainClass) {
+        return Optional.of(reader.getClassName().replace("/", "."));
       }
 
-    } catch (NotFoundException ex) {
-      // Thrown if 'java.lang.String' is not found in classPool.
-      throw new RuntimeException(ex);
+    } catch (ArrayIndexOutOfBoundsException ignored) {
```
[the remark this answers is not in the document]
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Having trouble finding any, but I'll keep looking and let you know.

[the remark this answers is not in the document]
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: I'd be curious to understand how this would trigger? Does the mean the `.class`es we are inspecting are potentially corrupt?

**Comment by loosebazooka** [comment]
Travis should catch any java11 incompatibilities?

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: This exception was thrown during one of our tests (`testFindMainClass_simple`, while trying to load `NotAClassButMightBe.class`). `ClassReader` does a bunch of stuff by indexing into a byte array without checking, and `NotAClassButMightBe.class` is just an empty file, so it just ends up going out of bounds.

**Comment by chanseokoh** [comment]
> Travis should catch any java11 incompatibilities?

What "compatibility" means may depend on the context. I don't think we have an integration test where the end-user project produces Java 9+ (including 11) classes and we verify we can find the main method in them. But I want to believe ASM is handling Java 11 classes well, as @TadCordle said.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Normally I would except these kind of exceptions to slip through directly to the user, but this would probably be a difficult error to debug on the user side. It's nice for us to detect these issues, but it might actually be okay to fail here. We don't want to be packaging bad class files anyway.

[the remark this answers is not in the document]
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: I don't think we should fail the build, I think it should behave the same way it would if it found a non-class file, but maybe we should log a warning instead of a debug message?

I don't even know if it's up to us to decide if their project files are valid or not, at least not at this stage where it's just trying to find the main class.

[the remark this answers is not in the document]
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Sounds good. Might make sense to move the catch to exactly the method we expect to throw it, so we are isolating for that single case.

[the remark this answers is not in the document]
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Just a final thing: It's not obviously clear to me which method call is throwing this exception as a developer, as it's not exposed by the API. Isolating it to the exact method call with a separate try block would clarify that. It does make the logic of this block a little more complicated though.

[the remark this answers is not in the document]
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Yeah, it's `ClassReader`. And I think it's already isolated to the degree it should be. It seems like all 3 places the `ClassReader` is used do some unchecked array indexing. I can add a comment though.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java [hunk 2]
```diff
@@ -106,6 +106,28 @@ public Type getType() {
     }
   }
 
+  /** {@link ClassVisitor} that keeps track of whether or not it has visited a main class. */
+  private static class MainClassVisitor extends ClassVisitor {
+
+    private boolean visitedMainClass;
+
+    private MainClassVisitor() {
+      super(Opcodes.ASM6);
```
[the remark this answers is not in the document]
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: I think I did this already, try refreshing?

[the remark this answers is not in the document]
Comment (briandealwis) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Oops!  Dunno how that happened.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java [hunk 3]
```diff
@@ -106,6 +107,29 @@ public Type getType() {
...
+
+    private MainClassVisitor() {
+      super(Opcodes.ASM7);
+    }
+
+    @Override
+    @Nullable
+    public MethodVisitor visitMethod(
+        int access, String name, String descriptor, String signature, String[] exceptions) {
+      if (access == (Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC)
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Nitpicky, but it seems `final` and `@Deprecated` seem to be valid for PSVMs.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java [hunk 4]
```diff
@@ -106,6 +107,29 @@ public Type getType() {
...
+          && name.equals("main")
+          && descriptor.equals(
+              Type.getMethodDescriptor(Type.VOID_TYPE, Type.getType(String[].class)))) {
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: Worth making this a constant?

**Comment by TadCordle** [comment]
FYI, tested with spring petclinic and it works :+1:

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java [hunk 5]
```diff
@@ -157,35 +187,21 @@ public Result find() {
...
-
-      } catch (IOException ex) {
-        // Could not read class file.
-        eventDispatcher.dispatch(LogEvent.warn("Could not read file: " + file));
+    MainClassVisitor mainClassVisitor = new MainClassVisitor();
+    try (InputStream classFileInputStream = Files.newInputStream(file)) {
+      ClassReader reader = new ClassReader(classFileInputStream);
+      reader.accept(mainClassVisitor, 0);
+      if (mainClassVisitor.visitedMainClass) {
+        return Optional.of(reader.getClassName().replace("/", "."));
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java]: tiny nit: could use char `'/', '.'`
